velox/dwio/dwrf/reader/SelectiveTimestampColumnReader.cpp

```diff
@@ -100,63 +115,73 @@ void SelectiveTimestampColumnReader::readHelper(RowSet rows) {
     decodeWithVisitor<velox::dwrf::RleDecoderV2<false>>(nano_.get(), visitor);
   }
   readOffset_ += numRows;
-}
 
-void SelectiveTimestampColumnReader::read(
-    vector_size_t offset,
-    RowSet rows,
-    const uint64_t* incomingNulls) {
-  prepareRead<int64_t>(offset, rows, incomingNulls);
-  VELOX_CHECK(!scanSpec_->filter());
-  bool isDense = rows.back() == rows.size() - 1;
-  if (isDense) {
-    readHelper<true>(rows);
-  } else {
-    readHelper<false>(rows);
-  }
-}
+  // Merge the seconds and nanos into 'values_'
+  auto secondsData = secondsValues_->as<int64_t>();
+  auto nanosData = values_->as<uint64_t>();
+  // If filter is present, return nulls in read to be processed.
+  auto rawNulls = nullsInReadRange_
+      ? ((returnReaderNulls_ || filter) ? nullsInReadRange_->as<uint64_t>()
+                                        : rawResultNulls_)
+      : nullptr;
+  auto tsValues = AlignedBuffer::allocate<Timestamp>(numValues_, &memoryPool_);
+  auto rawTs = tsValues->asMutable<Timestamp>();
 
-namespace {
-void fillTimestamps(
-    Timestamp* timestamps,
-    const uint64_t* nullsPtr,
-    const int64_t* secondsPtr,
-    const uint64_t* nanosPtr,
-    vector_size_t numValues) {
-  for (vector_size_t i = 0; i < numValues; i++) {
-    if (!nullsPtr || !bits::isBitNull(nullsPtr, i)) {
-      auto nanos = nanosPtr[i];
+  for (vector_size_t i = 0; i < numValues_; i++) {
+    if (!rawNulls || !bits::isBitNull(rawNulls, i)) {
+      auto nanos = nanosData[i];
       uint64_t zeros = nanos & 0x7;
       nanos >>= 3;
       if (zeros != 0) {
         for (uint64_t j = 0; j <= zeros; ++j) {
           nanos *= 10;
         }
       }
-      auto seconds = secondsPtr[i] + EPOCH_OFFSET;
+      auto seconds = secondsData[i] + EPOCH_OFFSET;
       if (seconds < 0 && nanos != 0) {
         seconds -= 1;
       }
-      timestamps[i] = Timestamp(seconds, nanos);
+      rawTs[i] = Timestamp(seconds, nanos);
     }
   }
+  values_ = tsValues;
+  rawValues_ = values_->asMutable<char>();
+
+  // Process filter if present.
+  switch (filter ? filter->kind() : common::FilterKind::kAlwaysTrue) {
+    case common::FilterKind::kAlwaysTrue:
+      break;
+    case common::FilterKind::kTimestampRange:
+      processFilter(filter);
+      break;
+    default:
+      VELOX_UNSUPPORTED("Unsupported filter.");
```

Comment:
Do we want to handle null filter here as well?

Reply:
Let me add that! Thanks for the review.